trap.go

```diff
@@ -443,11 +443,61 @@ func (t *TrapListener) debugTrapHandler(s *SnmpPacket, u *net.UDPAddr) {
 // UnmarshalTrap unpacks the SNMP Trap.
 //
 // NOTE: the trap code is currently unreliable when working with snmpv3 - pull requests welcome
-func (x *GoSNMP) UnmarshalTrap(trap []byte, useResponseSecurityParameters bool) (result *SnmpPacket, err error) {
-	result = new(SnmpPacket)
+func (x *GoSNMP) UnmarshalTrap(trap []byte, useResponseSecurityParameters bool) (*SnmpPacket, error) {
+	result := new(SnmpPacket)
+	// Get only the version from the header of the trap
+	_, err := x.unmarshalVersionFromHeader(trap, result)
+	if err != nil {
+		x.Logger.Printf("UnmarshalTrap version unmarshal: %s\n", err)
+		return nil, err
+	}
+	// If there are multiple users configured and the SNMP trap is v3, see which user has valid credentials
+	// by iterating through the list matching the identifier and seeing which credentials are authentic / can be used to decrypt
+	if x.SecurityParametersTable != nil && result.Version == Version3 {
+		identifier, err := x.getTrapIdentifier(trap)
+		if err != nil {
+			x.Logger.Printf("UnmarshalTrap V3 get trap identifier: %s\n", err)
+			return nil, err
+		}
+		secParamsList, err := x.SecurityParametersTable.Get(identifier)
+		if err != nil {
+			x.Logger.Printf("UnmarshalTrap V3 get security parameters from table: %s\n", err)
+			return nil, err
+		}
+		for _, secParams := range secParamsList {
+			// Copy the trap and re-initialize the packet with new security parameters to unmarshal with
+			cpTrap := make([]byte, len(trap))
+			copy(cpTrap, trap)
+			result = new(SnmpPacket)
+			e := secParams.InitSecurityKeys()
+			if e != nil {
+				return nil, err
+			}
+			result.SecurityParameters = secParams.Copy()
+			result, e = x.UnmarshalTrapBase(cpTrap, result, true)
+			if result != nil {
+				return result, e
+			}
+		}
+		return nil, fmt.Errorf("no credentials successfully unmarshaled trap")
```

Comment:
If `err` would already have been defined by the func statement, something like this could be done in order to show the last error for example:
```suggestion
		return nil, fmt.Errorf("no credentials successfully unmarshaled trap: %w", err)
```